substrate/primitives/virtualization/src/tests.rs

```diff
@@ -0,0 +1,290 @@
+	assert_eq!(ret, Err(ExecError::OutOfGas));
+	assert!(!state.exit);
+	assert_eq!(state.user.counter, 6_666);
+	assert_eq!(state.gas_left, 0);
+}
+
+/// Call same function with different gas limits and make sure they consume the same amount of gas.
+fn gas_consumption_works(program: &[u8]) {
+	let gas_limit_0 = GAS_MAX;
+	let gas_limit_1 = gas_limit_0 / 2;
+
+	let mut instance = Virt::instantiate(program).unwrap();
+	let mut state = SharedState {
+		gas_left: gas_limit_0,
+		exit: false,
+		user: State { counter: 0, memory: Some(instance.memory()), ..Default::default() },
+	};
+	let ret = instance.execute("counter", syscall_handler, &mut state);
+	let gas_consumed = gas_limit_0 - state.gas_left;
+	assert_eq!(ret, Ok(()));
+
+	let mut instance = Virt::instantiate(program).unwrap();
+	let mut state = SharedState {
+		gas_left: gas_limit_1,
+		exit: false,
+		user: State { counter: 0, memory: Some(instance.memory()), ..Default::default() },
+	};
+	let ret = instance.execute("counter", syscall_handler, &mut state);
+	assert_eq!(ret, Ok(()));
+	assert_eq!(gas_consumed, gas_limit_1 - state.gas_left);
+}
+
+/// Make sure that globals are reset for a new instance.
+fn memory_reset_on_instantiate(program: &[u8]) {
+	let mut instance = Virt::instantiate(program).unwrap();
+	let mut state = SharedState {
+		gas_left: GAS_MAX,
+		exit: false,
+		user: State { counter: 0, memory: Some(instance.memory()), ..Default::default() },
+	};
+	let ret = instance.execute("offset", syscall_handler, &mut state);
+	assert_eq!(ret, Ok(()));
+	assert_eq!(state.user.counter, 3);
+
+	let mut instance = Virt::instantiate(program).unwrap();
+	let ret = instance.execute("offset", syscall_handler, &mut state);
+	assert_eq!(ret, Ok(()));
+	assert_eq!(state.user.counter, 6);
+}
+
+/// Make sure globals are not reset between multiple calls into the same instance.
+fn memory_persistent(program: &[u8]) {
+	let mut instance = Virt::instantiate(program).unwrap();
+	let mut state = SharedState {
+		gas_left: GAS_MAX,
+		exit: false,
+		user: State { counter: 0, memory: Some(instance.memory()), ..Default::default() },
+	};
+	let ret = instance.execute("offset", syscall_handler, &mut state);
+	assert_eq!(ret, Ok(()));
+	assert_eq!(state.user.counter, 3);
+
+	let ret = instance.execute("offset", syscall_handler, &mut state);
+	assert_eq!(ret, Ok(()));
+	assert_eq!(state.user.counter, 7);
+}
+
+/// Calls a function that spawns another instance where it calls the `counter` entry point.
+fn counter_in_subcall(program: &[u8]) {
+	let mut instance = Virt::instantiate(program).unwrap();
+	let mut state = SharedState {
+		gas_left: GAS_MAX,
+		exit: false,
+		user: State { counter: 0, memory: Some(instance.memory()), program: program.to_vec() },
+	};
+	let ret = instance.execute("do_subcall", syscall_handler, &mut state);
+	assert_eq!(ret, Ok(()));
+	assert!(!state.exit);
+	// sub call should not affect parent state
+	assert_eq!(state.user.counter, 0);
```

Comment:
should we add a bool flag in the user state to at least verify that it indeed did execute the subcall